satellite/metabase/rangedloop/rangedlooptest/callbackobserver.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (C) 2022 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package rangedlooptest
+
+import (
+	"context"
+	"time"
+
+	"storj.io/storj/satellite/metabase/rangedloop"
+	"storj.io/storj/satellite/metabase/segmentloop"
+)
+
+var _ rangedloop.Observer = (*CallbackObserver)(nil)
+var _ rangedloop.Partial = (*CallbackObserver)(nil)
+
+// CallbackObserver can be used to easily attach logic to the ranged segment loop during tests.
+type CallbackObserver struct {
+	OnProcess func(context.Context, []segmentloop.Segment) error
+	OnStart   func(context.Context, time.Time) error
+	OnFork    func(context.Context) (rangedloop.Partial, error)
+	OnJoin    func(context.Context, rangedloop.Partial) error
+	OnFinish  func(context.Context) error
+}
+
+// NewCallbackObserver initializes a new CallbackObserver with no-op callbacks.
+func NewCallbackObserver() *CallbackObserver {
```

Comment:
Since the fields on the callback observer are exported, i think it is more robust to not assume these callbacks are non-nil at the call sites, i.e., it feels more flexible and robust to do "if" checks. When a user wants to set just one callback, it is more convenient to be able to construct it via a composite literal than have to call NewCallbackObserver and then overwrite a callback or provide it's own no-op callbacks when initializing with the composite literal.

Reply:
Done